src/voter/mod.rs

```diff
@@ -849,49 +947,109 @@ impl<H, N, E: Environment<H, N>, CommitIn, CommitOut> Voter<H, N, E, CommitIn, C
 		Ok(())
 	}
 
-	fn process_commits(&mut self) -> Result<(), E::Error> {
-		if let Async::Ready((round_number, _)) = self.committer.poll()? {
-			let prospective_round_number =
-				self.prospective_round.as_ref().map(|r| r.votes.number());
+	/// Process all incoming messages from other nodes.
+	///
+	/// Commit messages are handled with extra care. If a commit message references
+	/// a currently backgrounded round, we send it to that round so that when we commit
+	/// on that round, our commit message will be informed by those that we've seen.
+	///
+	/// Otherwise, we will simply handle the commit and issue a finalization command
+	/// to the environment.
+	///
+	/// When witnessing a commit from a future round, we also start a prospective round
+	/// that can allow us to catch up when we've been stuck behind.
+	fn process_incoming(&mut self) -> Result<(), E::Error> {
+		let mut highest_incoming_foreign_commit = None;
+		while let Async::Ready(Some(item)) = self.global_in.poll()? {
+			match item {
+				CommunicationIn::Commit(round_number, commit) => {
+					trace!(target: "afg", "Got commit for round_number {:?}: target_number: {:?}, target_hash: {:?}",
+						round_number,
+						commit.target_number,
+						commit.target_hash,
+					);
+
+					let commit: Commit<_, _, _, _> = commit.into();
+
+					// if the commit is for a background round dispatch to round committer.
+					// that returns Some if there wasn't one.
+					if let Some(commit) = self.past_rounds.import_commit(round_number, commit) {
+						// otherwise validate the commit and signal the finalized block
+						// (if any) to the environment
+						if let Some((finalized_hash, finalized_number)) = validate_commit(
+							&commit,
+							&self.voters,
+							&*self.env,
+						)? {
+							match highest_incoming_foreign_commit {
```

Comment:
Is this equal to 
```rust
highest_incoming_foreign_commit = highest_incoming_foreign_commit.map_or(round_number,
	|n| ::std::cmp::max(round_number, n));
```
?

Then I'd prefer this, because I feel it is more expressive of what is happening than the current match.